### PR TITLE
table,tablecodec,util/types: make it possible to read empty enum value

### PR DESCRIPTION
**Pull request by tiancaiamao**
Empty or NULL Enumeration Values is valid in mysql, https://dev.mysql.com/doc/refman/5.7/en/enum.html
this PR make it possible to read out the empty enum value.

```
set sql_mode='';
create table t (e enum('Y', 'N'));
insert into t values (0);

mysql> select * from t;
+------+
| e    |
+------+
|      |
+------+
1 row in set (0.00 sec)
```

PTAL @zimulala @hanfei1991 @shenli

### CODE REVIEW

**Comment by zimulala** [comment]
I think it's better to merge to rc3 rather than the master branch。

**Comment by tiancaiamao** [comment]
This is compatible with mysql's behavior, mysql also handle empty enum value. 
@zimulala

File: tablecodec/tablecodec.go [hunk 1]
```diff
@@ -416,10 +416,8 @@ func unflatten(datum types.Datum, ft *types.FieldType, loc *time.Location) (type
 		datum.SetValue(dur)
 		return datum, nil
 	case mysql.TypeEnum:
-		enum, err := types.ParseEnumValue(ft.Elems, datum.GetUint64())
-		if err != nil {
-			return datum, errors.Trace(err)
-		}
+		// ignore error deliberately, to read empty enum value.
+		enum, _ := types.ParseEnumValue(ft.Elems, datum.GetUint64())
```
Comment (shenli) [inline tablecodec/tablecodec.go]: Should we also update coprocessor?

**Comment by shenli** [comment]
LGTM
@AndreMouche Should we also update coprocessor?

File: tablecodec/tablecodec.go (hunk 1 above)
Comment (tiancaiamao) [inline tablecodec/tablecodec.go]: TiKV have no idea about enum type, is just a integer in TiKV, 
TiDB handle the convertion from integer to enum, in the `unflatten` function.

**Comment by zimulala** [comment]
But the MySQL has the warnings. 
```
mysql> set sql_mode='';
Query OK, 0 rows affected, 1 warning (0.00 sec)
mysql> create table t (e enum('Y', 'N'));
Query OK, 0 rows affected (0.03 sec)
mysql> insert into t value(0);
Query OK, 1 row affected, 1 warning (0.02 sec)
mysql> show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1265 | Data truncated for column 'e' at row 1 |
+---------+------+----------------------------------------+
1 row in set (0.00 sec)
```

File: util/types/datum.go [hunk 2]
```diff
@@ -1201,10 +1201,10 @@ func (d *Datum) convertToMysqlEnum(sc *variable.StatementContext, target *FieldT
 		e, err = ParseEnumValue(target.Elems, uintDatum.GetUint64())
 	}
 	if err != nil {
-		return invalidConv(d, target.Tp)
+		err = errors.Wrap(err, ErrTruncated)
 	}
 	ret.SetValue(e)
-	return ret, nil
+	return ret, err
```
Comment (zimulala) [inline util/types/datum.go]: Add error trace.

**Comment by zimulala** [comment]
Add more tests.

File: util/types/datum.go (hunk 2 above)
Comment (tiancaiamao) [inline util/types/datum.go]: errors.Wrap have the error trace info.

**Comment by tiancaiamao** [comment]
PTAL @zimulala

File: tablecodec/tablecodec.go [hunk 3]
```diff
@@ -416,10 +416,8 @@ func unflatten(datum types.Datum, ft *types.FieldType, loc *time.Location) (type
 		datum.SetValue(dur)
 		return datum, nil
 	case mysql.TypeEnum:
-		enum, err := types.ParseEnumValue(ft.Elems, datum.GetUint64())
-		if err != nil {
-			return datum, errors.Trace(err)
-		}
+		// ignore error deliberately, to read empty enum value.
```
Comment (zimulala) [inline tablecodec/tablecodec.go]: s/ignore/Ignore

File: executor/executor_test.go [hunk 4]
```diff
@@ -1863,3 +1863,21 @@ func (s *testSuite) TestFuncREPEAT(c *C) {
...
 	r.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil> <nil>"))
 }
+
+func (s *testSuite) TestEmptyEnum(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	defer func() {
+		s.cleanEnv(c)
+		testleak.AfterTest(c)()
+	}()
+	tk.MustExec("use test")
```
Comment (zimulala) [inline executor/executor_test.go]: Add test about `sql_mode = "STRICT_TRANS_TABLES"`.

**Comment by zimulala** [comment]
Rest LGTM

File: executor/executor_test.go (hunk 4 above)
Comment (tiancaiamao) [inline executor/executor_test.go]: Done. @zimulala

**Comment by zimulala** [comment]
LGTM

**Pull request merged**